frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -0,0 +1,206 @@
+				err = dispatchWorkdir(d, c)
+			case *instructions.AddCommand:
+				err = dispatchCopy(d, c.SourcesAndDest, llb.Local("context"))
+			case *instructions.CopyCommand:
+				l := llb.Local("context")
+				if c.From != "" {
+					index, err := strconv.Atoi(c.From)
+					if err != nil {
+						stn, ok := stagesByName[strings.ToLower(c.From)]
+						if !ok {
+							return nil, errors.Errorf("stage %s not found", c.From)
+						}
+						l = stn.state
+					} else {
+						if index >= len(allStages) {
+							return nil, errors.Errorf("invalid stage index %d", index)
+						}
+						l = allStages[index].state
+					}
+				}
+				err = dispatchCopy(d, c.SourcesAndDest, l)
+			default:
+				continue
+			}
+			if err != nil {
+				return nil, err
+			}
+		}
+
+	}
+
+	if target == "" {
+		return &allStages[len(allStages)-1].state, nil
+	}
+
+	state, ok := stagesByName[target]
+	if !ok {
+		return nil, errors.Errorf("target stage %s could not be found", target)
+	}
+	return &state.state, nil
+}
+
+type dispatchState struct {
+	state llb.State
+	image *ocispec.Image
+	stage instructions.Stage
+}
+
+func dispatchEnv(d *dispatchState, c *instructions.EnvCommand) error {
+	for _, e := range c.Env {
+		d.state = d.state.AddEnv(e.Key, e.Value)
+		// addEnv(d.image, e.Key, e.Value)
+	}
+	return nil
+}
+
+func dispatchRun(d *dispatchState, c *instructions.RunCommand) error {
+	var args []string = c.CmdLine
+	if c.PrependShell {
+		args = append(defaultShell(), strings.Join(args, " "))
+	}
+	d.state = d.state.Run(llb.Args(args)).Root()
+	return nil
+}
+
+func dispatchWorkdir(d *dispatchState, c *instructions.WorkdirCommand) error {
+	d.state = d.state.Dir(c.Path)
+	wd := c.Path
+	if !path.IsAbs(c.Path) {
+		wd = path.Join("/", d.image.Config.WorkingDir, wd)
+	}
+	d.image.Config.WorkingDir = wd
+	return nil
+}
+
+func dispatchCopy(d *dispatchState, c instructions.SourcesAndDest, sourceState llb.State) error {
+	// TODO: this should use CopyOp instead. Current implementation is inefficient and doesn't match Dockerfile path suffixes rules
+	img := llb.Image("docker.io/library/alpine@sha256:1072e499f3f655a032e88542330cf75b02e7bdf673278f701d7ba61629ee3ebe")
+
+	args := []string{"cp", "-a"}
+	for _, src := range c.Sources() {
+		args = append(args, filepath.Join("/src", src))
+	}
+	args = append(args, filepath.Join("/dest", c.Dest()))
+
+	run := img.Run(llb.Args(args))
+	run.AddMount("/src", sourceState)
+	d.state = run.AddMount("/dest", d.state)
+	return nil
+}
+
+func dispatchMaintainer(d *dispatchState, c instructions.MaintainerCommand) error {
+	d.image.Author = c.Maintainer
+	return nil
+}
+
+func dispatchLabel(d *dispatchState, c instructions.LabelCommand) error {
+	if d.image.Config.Labels == nil {
+		d.image.Config.Labels = make(map[string]string)
+	}
+	for _, v := range c.Labels {
+		d.image.Config.Labels[v.Key] = v.Value
+	}
+	return nil
+}
+
+// func dispatchOnbuild(d *dispatchState, c *instructions.OnbuildCommand) error {
```

Comment:
Please open GitHub issues for ONBUILD and COPY?

Reply:
#108 #109